src/Processors/QueryPlan/ReadFromMergeTree.cpp

```diff
@@ -1262,12 +1256,15 @@ MergeTreeDataSelectAnalysisResultPtr ReadFromMergeTree::selectRangesToReadImpl(
     return std::make_shared<MergeTreeDataSelectAnalysisResult>(MergeTreeDataSelectAnalysisResult{.result = std::move(result)});
 }
 
-void ReadFromMergeTree::requestReadingInOrder(size_t prefix_size, int direction, size_t limit)
+bool ReadFromMergeTree::requestReadingInOrder(size_t prefix_size, int direction, size_t limit)
 {
     /// if dirction is not set, use current one
     if (!direction)
         direction = getSortDirection();
 
+    if (direction != 1 && isFinal(query_info))
```

Comment:
```suggestion
    /// Disable read-in-order optimization for reverse order with final. 
    /// Otherwise, it can lead to incorrect final behavior because the implementation may rely on the reading in direct order). 
    if (direction != 1 && isFinal(query_info))
```